[nnc] Added a new API to compress all buffers in a given statement

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61293
* **#61087**
* #61085

Fixes #60991 

Differential Revision: [D29506677](https://our.internmc.facebook.com/intern/diff/D29506677)

navahgar committed:
[nnc] Added a new API to compress all buffers in a given statement

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29506677).

navahgar committed:
Update on "[nnc] Added a new API to compress all buffers in a given statement"


Fixes #60991 

Differential Revision: [D29506677](https://our.internmc.facebook.com/intern/diff/D29506677)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29506677).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.h:
```diff
@@ -412,6 +417,13 @@ class TORCH_API LoopNest {
 
+  // Compresses all buffers in the given statement.
+  //
+  // NOTE: This API assumes that there are no accesses to buffers outside
+  // the given statement. So, this should be called with the entire
+  // kernel statement to avoid incorrect buffer compressions.
```
It might be a good idea to add a check to IR verifier that would detect invalidly compressed buffers (not for this PR, but might be worth mentioning as a TODO).

navahgar replied:
Added a TODO

ZolotukhinM commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -6976,5 +6976,75 @@ TEST(LoopNest, compressBufferIndicesMixed) {
+  torch::jit::testing::FileCheck().run(verification_pattern, oss.str());
+
+  ASSERT_EQ(A->ndim(), 2);
+  ASSERT_EQ(dynamic_cast<const IntImm*>(A->dim(0))->value(), 1);
```
Might be worth reusing `IS_IMM_WITH_VAL` macros defined in `test_simplify.cpp`, which can be used like this:
https://github.com/pytorch/pytorch/blob/dec5aa2260cef540b622bd9a9504b6f11cb1f607/test/cpp/tensorexpr/test_simplify.cpp#L665

navahgar replied:
Good point. I need to refactor that macro out of `test_simplify` first. Also, other tests above also use this. So, I will do it as a follow up.

ZolotukhinM commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -6976,5 +6976,75 @@ TEST(LoopNest, compressBufferIndicesMixed) {
+  //     C[i,m] = B[i,m]
+  //   }
+  // }
+  Buf* A = new Buf("A", {new IntImm(100), new IntImm(200)}, kInt);
+  Buf* B = new Buf("B", {new IntImm(100), new IntImm(200)}, kInt);
+  Buf* C = new Buf("C", {new IntImm(100), new IntImm(200)}, kInt);
+  BufHandle a_buf(A);
+  BufHandle b_buf(B);
+  BufHandle c_buf(C);
```
We probably can create just `BufHandle`s - it seems that we use raw buffers only in the final checks, but that can be done through handles as well.

navahgar replied:
Ok, sure. Since there are tests above that does the same, I will do this as a follow up.

ZolotukhinM reviewed (approved):
Looks good!

navahgar committed:
Update on "[nnc] Added a new API to compress all buffers in a given statement"


Fixes #60991 

Differential Revision: [D29506677](https://our.internmc.facebook.com/intern/diff/D29506677)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29506677).

navahgar committed:
Update on "[nnc] Added a new API to compress all buffers in a given statement"


Fixes #60991 

Differential Revision: [D29506677](https://our.internmc.facebook.com/intern/diff/D29506677)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29506677).
